src/mdlComponentHandler.js

```diff
@@ -99,37 +86,44 @@ var componentHandler = (function() {
    * the element to.
    */
   function upgradeElementInternal(element, optJsClass) {
-    // Only upgrade elements that have not already been upgraded.
+    // Verify argument type.
+    if (!(typeof element === 'object' && element instanceof Element)) {
+      throw new Error('Invalid argument provided to upgrade MDL element.');
+    }
     var dataUpgraded = element.getAttribute('data-upgraded');
-
-    var registeredClasses = [];
+    // Use `['']` as default value to conform the `,name,name...` style.
+    var upgradedList = dataUpgraded === null ? [''] :
+                       String.prototype.split.call(dataUpgraded, ',');
+    var classesToUpgrade = [];
     // If jsClass is not provided scan the registered components to find the
     // ones matching the element's CSS classList.
     if (!optJsClass) {
-      registeredClasses = registeredComponents_.filter(function(component) {
-        return element.classList.contains(component.cssClass) &&
-          !isElementUpgraded_(element, component.className);
+      var classList = element.classList;
+      registeredComponents_.forEach(function (component) {
+        // Match CSS & Not upgraded & Not to be upgraded.
+        if (classList.contains(component.cssClass) &&
+            upgradedList.indexOf(component.className) === -1 &&
+            classesToUpgrade.indexOf(component) === -1) {
+          classesToUpgrade.push(component);
+        }
       });
-    } else if (!isElementUpgraded_(element, optJsClass)) {
-      registeredClasses.push(findRegisteredClass_(optJsClass));
+    } else if (upgradedList.indexOf(optJsClass) === -1) {
+      classesToUpgrade.push(findRegisteredClass_(optJsClass));
     }
 
     // Upgrade the element for each classes.
-    for (var i = 0, l = registeredClasses.length; i < l; i++) {
-      var registeredClass = registeredClasses[i];
+    for (var i = 0, n = classesToUpgrade.length, registeredClass; i < n; i++) {
+      registeredClass = classesToUpgrade[i];
       if (registeredClass) {
         // Mark element as upgraded.
-        if (dataUpgraded === null) {
-          dataUpgraded = '';
-        }
-        element.setAttribute('data-upgraded', dataUpgraded + ',' +
-          registeredClass.className);
+        upgradedList.push(registeredClass.className);
+        element.setAttribute('data-upgraded', upgradedList.join(','));
```

Comment:
I don't think there is a major performance impact nor buggy behavior with the current implementation. It handles the "if one of many fails" cases well. So it lgtm unless there is a very compelling reason otherwise.